Add a --mode flag to :hints

haasn opened the pull request:
This allows a specific keybinding, for whatever reason, to override the
default mode. Examples of when this could be useful:

* :hint --rapid --mode=word  (to type them more rapidly)
* :hint --mode=letter input  (if the default mode is number)

Also reword the description of 'group' to make the distinction between
'group' and 'mode' clearer.

haasn commented:
Incidentally, isn't there a way to avoid the annoying duplication of documentation between the asciidoc and the implementation comments?

The-Compiler commented:
> Incidentally, isn't there a way to avoid the annoying duplication of documentation between the asciidoc and the implementation comments?

The files are autogenerated by running `scripts/dev/src2asciidoc.py` :wink:

haasn commented:
> The files are autogenerated by running `scripts/dev/src2asciidoc.py`

D'oh. Updated with that script used to generate the documentation instead. (And some minor tweaks to account for this)

The-Compiler commented:
> D'oh. Updated with that script used to generate the documentation instead.

I added a comment to the files in ec21041516d3edf0f1512b92131b45a40248a4be to prevent this to happen in the future :wink:

The-Compiler commented:
Can you please add a test for this to `tests/end2end/features/hint.feature`? Should be relatively easy by setting the mode, then using `:hint` with `--mode`, using `:follow-hint` and checking if the right hint was followed.

The-Compiler commented on qutebrowser/browser/hints.py:
```diff
@@ -732,11 +739,15 @@ def start(self, rapid=False, group=webelem.Group.all, target=Target.normal,
         self._context.tab = tab
         self._context.target = target
         self._context.rapid = rapid
+        self._context.hint_mode = mode
```
Please also set `self.hint_mode = None` in `HintContext.__init__`

haasn commented:
Updated to address the mentioned issues and failures, added a test, and fixed a bug uncovered by the test case. (Not sure if the fix is the most elegant way to solve this - I basically added a function to check what the currently active hinting mode is)

The-Compiler commented on qutebrowser/browser/hints.py:
```diff
@@ -748,6 +760,13 @@ def start(self, rapid=False, group=webelem.Group.all, target=Target.normal,
+        if not self._context or not self._context.hint_mode:
+            return None
+
+        return self._context.hint_mode
```
I think you could simplify this a bit, and make the check for `None` more explicit:

```python
if self._context is None:
    return None
return self._context.hint_mode
```

On the other hand, will this ever be called when `self._context` is `None`? It only gets called by `HintKeyParser` (i.e., in hint mode), so `self._context` should always be set, no?

haasn replied:
Good suggestion. Is the lack of a newline between the `if` body and the `return` a python stylistic thing? Looks weird to me

> It only gets called by HintKeyParser (i.e., in hint mode), so self._context should always be set, no?

I don't see a need for this function to make that assumption. Who knows what might use it in the future?

The-Compiler replied:
I usually don't use newlines before a `return` if the function is short enough, but it's probably purely a personal preference - no strong feelings either way.

> An explicit check is more robust and only costs one line.

Fair enough!

haasn commented:
Simplified the check as discussed. (But I kept the check for safety, and the newline for readability)

(Incidentally, is there a way to re-run the coverage check? I believe it should now no longer decrease coverage)

The-Compiler commented:
> Incidentally, is there a way to re-run the coverage check? I believe it should now no longer decrease coverage

The comment above is updated when a build is finished. And indeed it doesn't.

The-Compiler commented:
Thanks!

The-Compiler commented on qutebrowser/browser/hints.py:
```diff
@@ -748,6 +760,13 @@ def start(self, rapid=False, group=webelem.Group.all, target=Target.normal,
                                             only_visible=True)
 
+    def current_mode(self):
+        """ Returns the currently active hinting mode (or None otherwise). """
```
I removed the spaces inside the docstring here - now that one is a common [python convention](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings).

lahwaacz commented on qutebrowser/keyinput/modeparsers.py:
```diff
@@ -189,7 +189,7 @@ def _handle_special_key(self, e):
             else:
                 return super()._handle_special_key(e)
-        elif config.get('hints', 'mode') != 'number':
+        elif hintmanager.current_mode() != 'number':
```
This should have been `hintmanager._context.hint_mode`.

lahwaacz replied:
And IMO it should not be part of context - it's a config value that rarely changes.

The-Compiler replied:
I disagree with both :wink: Doing `hintmanager._context.hint_mode` would mean accessing a "private" attribute (and making the whole context public isn't nice either).

As for the latter, the whole point of this PR is adding `--mode` to `:hint` to override the mode on a case-by-case basis :wink:

lahwaacz replied:
Right, I need more coffee :disappointed: I missed the `current_mode` getter because the `HintManager` itself does not use it - which would be much nicer. It could be even decorated with `@property`.

The-Compiler replied:
I wish we could use `@property` more, but with PyQt, if any exception happens inside a property handler, you get an `AttributeError` for the property instead, which is really confusing - that's why I'm trying to avoid it.
